### PR TITLE
Kubelet: Refactor all but image related functions in DockerInterface

### PR DESCRIPTION
**Pull request by Random-Liu**
For #23563.
Based on #23699 and #23844.

Only last 3 commits are new. This PR refactored all functions except image related functions, including:
- CreateExec
- StartExec
- InspectExec
- AttachToContainer
- Logs
- Info
- Version

@kubernetes/sig-node

### CODE REVIEW

**Comment by Random-Liu** [comment]
@k8s-bot node e2e test this issue #IGNORE.

**Comment by ncdc** [comment]
@Random-Liu FYI, in case you run into any weird behavior in exec or attach, fyi: https://github.com/kubernetes/kubernetes/blob/1032067ff98ec94515116c5fef9f108541486771/pkg/client/unversioned/remotecommand/v2.go#L124

**Comment by Random-Liu** [comment]
@ncdc Cool, thanks a lot! :)

**Comment by Random-Liu** [comment]
@k8s-bot test this issue #23952

**Comment by Random-Liu** [comment]
@k8s-bot test this issue #23772

**Comment by Random-Liu** [comment]
@yujuhong PTAL when you have time~ :)

File: pkg/kubelet/dockertools/docker.go [hunk 1]
```diff
@@ -71,9 +71,9 @@ type DockerInterface interface {
 	Logs(opts docker.LogsOptions) error
 	Version() (*docker.Env, error)
 	Info() (*docker.Env, error)
-	CreateExec(docker.CreateExecOptions) (*docker.Exec, error)
```
Comment (yujuhong) [inline pkg/kubelet/dockertools/docker.go]: commit LGTM.

File: pkg/kubelet/dockertools/manager.go [hunk 2]
```diff
@@ -1096,19 +1095,20 @@ func (dm *DockerManager) ExecInContainer(containerID kubecontainer.ContainerID,
 }
 
 func (dm *DockerManager) AttachContainer(containerID kubecontainer.ContainerID, stdin io.Reader, stdout, stderr io.WriteCloser, tty bool) error {
-	opts := docker.AttachToContainerOptions{
-		Container:    containerID.ID,
+	// TODO(random-liu): Do we really use the *Logs* field here?
```
Comment (yujuhong) [inline pkg/kubelet/dockertools/manager.go]: I don't see where go-dockerclient uses the "Logs" option at all. Did I miss anything?

File: pkg/kubelet/dockertools/manager.go [hunk 3]
```diff
@@ -873,101 +871,70 @@ func getDockerNetworkMode(container *dockertypes.ContainerJSON) string {
...
 // dockerVersion implementes kubecontainer.Version interface by implementing
-// Compare() and String() (which is implemented by the underlying semver.Version)
-// TODO: this code is the same as rktVersion and may make sense to be moved to
-// somewhere shared.
-type dockerVersion struct {
-	*semver.Version
-}
-
-// Older versions of Docker could return non-semantically versioned values (distros like Fedora
-// included partial values such as 1.8.1.fc21 which is not semver). Force those values to be semver.
```
Comment (yujuhong) [inline pkg/kubelet/dockertools/manager.go]: Do we have a test case for this?

File: pkg/kubelet/dockertools/manager_test.go [hunk 4]
```diff
@@ -1822,21 +1824,20 @@ func TestCheckVersionCompatibility(t *testing.T) {
 		// Invalid apiversion
 		{"invalid_api_version", false},
 	}
-	for i := range apiVersion {
-		apiVersion[i]++
+	for i := range intSlice {
```
Comment (yujuhong) [inline pkg/kubelet/dockertools/manager_test.go]: I have mixed feelings about this test. I know the test tries to be generic, but I find it harder to parse/read. IMO, it's easier to reason with with test cases such as "1.9.1", etc. I know it means that every time we bump the docker version, we'll have to manually change this test, but honestly, I think that's trivial enough and doesn't happen often (we bump docker version not more than twice per release). @Random-Liu, what to do you think?

File: pkg/kubelet/dockertools/docker.go (hunk 1 above)
Comment (MHBauer) [inline pkg/kubelet/dockertools/docker.go]: I concur

File: pkg/kubelet/dockertools/kube_docker_client.go [hunk 5]
```diff
@@ -367,6 +350,14 @@ func parseDockerTimestamp(s string) (time.Time, error) {
 	return time.Parse(time.RFC3339Nano, s)
 }
 
+// StreamOptions are the options used to configure the stream redirection
+type StreamOptions struct {
```
Comment (MHBauer) [inline pkg/kubelet/dockertools/kube_docker_client.go]: This is going to be a key struct.

File: pkg/kubelet/dockertools/manager_test.go (hunk 4 above)
Comment (Random-Liu) [inline pkg/kubelet/dockertools/manager_test.go]: Hahaha, the test is really ugly, I agree, :P. Maybe we can use a very old version and a very new version instead...Such as "0.0.0" and "9999.9999.9999", Hahaha~~WDYT?

File: pkg/kubelet/dockertools/manager.go (hunk 3 above)
Comment (Random-Liu) [inline pkg/kubelet/dockertools/manager.go]: Yeah, I removed it. https://github.com/kubernetes/kubernetes/pull/23907/commits/e64167b8940901e1e6b61dd313e3576c56fca587#diff-e2fb6215bbaa9210484ba6b85101007aL134

File: pkg/kubelet/dockertools/manager.go (hunk 2 above)
Comment (Random-Liu) [inline pkg/kubelet/dockertools/manager.go]: `Logs` is an argument in docker remote api https://docs.docker.com/engine/reference/api/docker_remote_api_v1.22/#attach-to-a-container, go-dockerclient only needs to accept it and encoding it into json.

File: pkg/kubelet/dockertools/manager_test.go (hunk 4 above)
Comment (yujuhong) [inline pkg/kubelet/dockertools/manager_test.go]: I think listing some explicit test cases is enough. This test makes me want to write a test to validate it ...

Comment (Random-Liu) [inline pkg/kubelet/dockertools/manager_test.go]: @yujuhong Hmm, test for test, good idea (kidding). I'll address the comment. :)

**Comment by Random-Liu** [comment]
@yujuhong @MHBauer Addressed comments, really thank you for reviewing! :)

**Comment by ingvagabund** [comment]
Too late but https://github.com/kubernetes/kubernetes/pull/20020. The semantic version is again not supported.

**Comment by ingvagabund** [comment]
Based on the code it should not work. Still, changing the version to semversion, TestVersion pass. Strange. @Random-Liu can you point me to a code where the semver parsing is actually carried?

**Comment by ingvagabund** [comment]
Whoo, it does not test the comparison. It just tests the ``Version`` and ``APIVersion`` are propagated through the docker manager.

**Pull request merged**